### PR TITLE
Handle errors in subprocesses better

### LINKED ISSUE #1112: Graceful parallel start up and tear down
* [x] Report error if fails to start workers
* [x] Automatically `load_all()` if working on test that?
* [x] Handle failures in the helper, startup, and teardown files.
* [x] Handle process cleanup on OSes that ps does not support.

### PR DESCRIPTION
**Pull request by gaborcsardi**
* [x] Report error if fails to start workers
* [x] Automatically `load_all()` if working on test that?
* [x] Handle failures in the helper, startup, and teardown files.
* [x] Handle process cleanup on OSes that ps does not support.

This is on top of #1141.

Closes #1112.

### CODE REVIEW

**Comment by gaborcsardi** [comment]
Hmmm, I still need to fix this from the CI on Linux:
```
-- FAILURE (test-parallel-setup.R:10:3): error in parallel setup code ----------
`err` inherits from `c_error/rlib_error/error/condition` not `testthat_process_error`.

-- FAILURE (test-parallel-setup.R:11:3): error in parallel setup code ----------
err$message does not match "Error in setup".
Actual value: "Invalid (uninitialized or closed?) connection object @processx-connection.c:1151 (processx__connection_find_chars)"
```

**Comment by gaborcsardi** [comment]
The CI failure is because the `install.packages()` bug, CRAN does not have a vctrs binary.

**Comment by gaborcsardi** [comment]
@hadley I think you can take a look at this, I'll publish processx (submitted) and callr (goes after processx) to get rid of the `r-lib/callr` remote.

File: R/parallel-taskq.R [hunk 1]
```diff
@@ -145,6 +139,32 @@ task_q <- R6::R6Class(
                                          private$tasks$args[[i]])
         }
       })
+    },
+
+    handle_error = function(msg, task_no) {
+      message("\n") # get out of the progress bar, if any
```
[the remark this answers is not in the document]
Comment (gaborcsardi) [reply R/parallel-taskq.R:145]: Sure.

File: R/test-files.R [hunk 2]
```diff
@@ -79,7 +79,11 @@ test_dir <- function(path,
     lifecycle::deprecate_warn("3.0.0", "test_dir(wrap = )")
   }
 
-  want_parallel <- find_parallel(path, package) && !is_parallel()
+  want_parallel <- find_parallel(
```
[the remark this answers is not in the document]
Comment (gaborcsardi) [reply R/test-files.R:82]: I put this logic into `find_parallel()`, so this is simpler now.

File: tests/testthat/test-parallel-setup.R [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
+test_that("error in parallel setup code", {
```
[the remark this answers is not in the document]
Comment (gaborcsardi) [reply tests/testthat/test-parallel-setup.R:2]: So we can run them in parallel? :)

File: tests/testthat/test-parallel-outside.R [hunk 4]
```diff
@@ -0,0 +1,12 @@
+
+test_that("error outside of test_that()", {
+  err <- tryCatch(
```
[the remark this answers is not in the document]
Comment (gaborcsardi) [reply tests/testthat/test-parallel-outside.R:3]: I could, but it is much easier to debug it manually if you can inspect the error object that you get, when you don't get the one you expected.

[the remark this answers is not in the document]
Comment (hadley) [reply tests/testthat/test-parallel-outside.R:3]: Ah that makes sense

File: tests/testthat/test-parallel-setup.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (hadley) [reply tests/testthat/test-parallel-setup.R:2]: Are they particularly slow?

[the remark this answers is not in the document]
Comment (gaborcsardi) [reply tests/testthat/test-parallel-setup.R:2]:
```
✔ |   2       | parallel-crash [2.5 s]
✔ |   1       | parallel-outside [1.7 s]
✔ |   2       | parallel-setup [1.0 s]
✔ |   2       | parallel-startup [0.7 s]
✔ |   0     1 | parallel-teardown
────────────────────────────────────────────────────────────────────────────────
Skip (test-parallel-teardown.R:3:3): teardown error
Reason: teardown errors are ignored
────────────────────────────────────────────────────────────────────────────────
✔ |   7       | parallel [2.5 s]
```

I can put them in one file of you want. 

FWIW I think parallel tests within parallel tests work now, at least on macOS, so we could run testthat's tests in parallel.

**Comment by gaborcsardi** [comment]
@hadley please take a look the three new commits.

**Comment by gaborcsardi** [comment]
Rebased on master.

**Review by hadley — COMMENTED** [review]
LGTM!

**Comment by gaborcsardi** [comment]
> LGTM!

Thanks! Let me add the env var that avoids the parallel/parallel nesting, it is simplest to do it here. Plus I also need to look at what's going on with the windows errors.

**Comment by gaborcsardi** [comment]
Added the env var to opt out, windows checks are fine, so ready to merge, I think.

**Comment by gaborcsardi** [comment]
Investigating why covr fails.

**Comment by gaborcsardi** [comment]
OK, all good now.

**Review by hadley — APPROVED** [review]

**Pull request merged**